dotCMS/src/main/java/com/dotcms/rendering/velocity/services/PageContextBuilder.java

```diff
@@ -0,0 +1,294 @@
+        boolean hasRemotePublishPermOverHTMLPage =
+                hasPublishPermOverHTMLPage && LicenseUtil.getLevel() >= LicenseLevel.STANDARD.level;
+        boolean hasEndPoints = UtilMethods.isSet(receivingEndpoints) && !receivingEndpoints.isEmpty();
+        boolean canUserWriteOnTemplate =
+                permissionAPI.doesUserHavePermission(template, PERMISSION_WRITE, user) && APILocator.getPortletAPI()
+                    .hasTemplateManagerRights(user);
+
+        ctxMap.put("dotPageMode", mode.name());
+
+        ctxMap.put("ADD_CHILDREN_HTMLPAGE_PERMISSION", hasAddChildrenPermOverHTMLPage);
+        ctxMap.put("EDIT_HTMLPAGE_PERMISSION", hasWritePermOverHTMLPage);
+        ctxMap.put("PUBLISH_HTMLPAGE_PERMISSION", hasPublishPermOverHTMLPage);
+        ctxMap.put("REMOTE_PUBLISH_HTMLPAGE_PERMISSION", hasRemotePublishPermOverHTMLPage);
+        ctxMap.put("REMOTE_PUBLISH_END_POINTS", hasEndPoints);
+        ctxMap.put("canAddForm", LicenseUtil.getLevel() >= LicenseLevel.STANDARD.level ? true : false);
+        ctxMap.put("canViewDiff", LicenseUtil.getLevel() >= LicenseLevel.STANDARD.level ? true : false);
+        ctxMap.put("pageChannel", pageChannel);
+        ctxMap.put("HTMLPAGE_ASSET_STRUCTURE_TYPE", true);
+        ctxMap.put("HTMLPAGE_IS_CONTENT", true);
+
+
+        ctxMap.put("EDIT_TEMPLATE_PERMISSION", canUserWriteOnTemplate);
+
+
+        ctxMap.put(ContainerLoader.SHOW_PRE_POST_LOOP, true);
+        ctxMap.put("HTMLPAGE_INODE", htmlPage.getInode());
+        ctxMap.put("HTMLPAGE_IDENTIFIER", htmlPage.getIdentifier());
+        ctxMap.put("HTMLPAGE_FRIENDLY_NAME", UtilMethods.espaceForVelocity(htmlPage.getFriendlyName()));
+        ctxMap.put("HTMLPAGE_TITLE", UtilMethods.espaceForVelocity(htmlPage.getTitle()));
+        ctxMap.put("TEMPLATE_INODE", template.getIdentifier());
+        ctxMap.put("HTMLPAGE_META", UtilMethods.espaceForVelocity(htmlPage.getMetadata()));
+        ctxMap.put("HTMLPAGE_DESCRIPTION", UtilMethods.espaceForVelocity(htmlPage.getSeoDescription()));
+        ctxMap.put("HTMLPAGE_KEYWORDS", UtilMethods.espaceForVelocity(htmlPage.getSeoKeywords()));
+        ctxMap.put("HTMLPAGE_SECURE", htmlPage.isHttpsRequired());
+        ctxMap.put("VTLSERVLET_URI", UtilMethods.encodeURIComponent(pageIdent.getURI()));
+        ctxMap.put("HTMLPAGE_REDIRECT", UtilMethods.espaceForVelocity(htmlPage.getRedirect()));
+        ctxMap.put("pageTitle", UtilMethods.espaceForVelocity(htmlPage.getTitle()));
+        ctxMap.put("friendlyName", UtilMethods.espaceForVelocity(htmlPage.getFriendlyName()));
+        ctxMap.put("HTML_PAGE_LAST_MOD_DATE", UtilMethods.espaceForVelocity(htmlPage.getFriendlyName()));
+        ctxMap.put("HTMLPAGE_MOD_DATE", new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS").format(htmlPage.getModDate()));
+
+
+
+    }
+
+
+
+    private void populateContainers() throws DotDataException, DotSecurityException {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 64 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)